### PR TITLE
add marshal `corrupt test suite` 

### PR DESCRIPTION
**Pull request by illia-li**
Changes:
1. Added `corrupt test suite`  for testing `marshal`, `unmarshal` functions.
2. Added `corrupt test` for `smallint`
3. Joined `serialization`, `utils` packages to `internal/tests/serialization`

### CODE REVIEW

File: internal/tests/serialization/set.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+						unmarshalInZero := newRefToZero(val)
+						s.runUnmarshalTest("unmarshal**zero", t, unmarshal, val, unmarshalInZero)
+					}
+				}
+			})
+		}
+	})
+}
+
+func (s Set) RunCorruptTest(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
```
Comment (dkropachev) [inline internal/tests/serialization/set.go:57]: Since you introduced a new method, now you need to explain what is the purpose of `RunCorruptTest`, outlining difference between `RunTest`

File: internal/tests/serialization/set.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+type Set struct {
+	Data   []byte
+	Values []interface{}
+
+	BrokenMarshalTypes   []reflect.Type
+	BrokenUnmarshalTypes []reflect.Type
+}
+
+func (s Set) RunTest(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
+	if name == "" {
```
Comment (dkropachev) [inline internal/tests/serialization/set.go:25]: Since you introduced a new method, now you need to explain what is the purpose of `RunTest`, outlining difference between `RunCorruptTest`

File: internal/tests/serialization/set.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+			}
+		}()
+		return f(val)
+	}()
+
+	if isTypeOf(val, s.BrokenMarshalTypes) {
+		if err != nil {
+			t.Fatalf("expected to not fail for (%T), but did fail", val)
+		}
+		t.Skipf("skipped bacause there is unsolved problem")
```
Comment (dkropachev) [inline internal/tests/serialization/set.go:174]: Do I see it correctly that `BrokenMarshalTypes` for `RunCurruptTest` means that test does not fail for given type ?

File: marshal_3_smallint_corrupt_test.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
+package gocql
```
[the remark this answers is not in the document]
Comment (illia-li) [reply marshal_3_smallint_corrupt_test.go]: Done

File: marshal_3_smallint_test.go [hunk 5]
```diff
@@ -3,9 +3,8 @@ package gocql
...
 	"testing"
 
-	"github.com/gocql/gocql/internal/tests/utils"
-	"github.com/gocql/gocql/marshal/tests/mod"
-	"github.com/gocql/gocql/marshal/tests/serialization"
+	"github.com/gocql/gocql/internal/tests/serialization"
+	"github.com/gocql/gocql/internal/tests/serialization/mod"
 )
 
 func TestMarshalSmallint(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (illia-li) [reply marshal_3_smallint_test.go:11]: Done

File: internal/tests/serialization/set.go (hunk 2 above)
Comment (illia-li) [reply internal/tests/serialization/set.go:25]: Done

File: internal/tests/serialization/set.go (hunk 3 above)
Comment (illia-li) [reply internal/tests/serialization/set.go:174]: It seems that I got confused myself when I wrote this code and messages.
I've redone the code and messages for more clarity.

File: internal/tests/serialization/set.go (hunk 1 above)
Comment (illia-li) [reply internal/tests/serialization/set.go:58]: Done

**Comment by sylwiaszunejko** [comment]
sorry for being so picky about commit messages but [move all code to internal/tests/serialization](https://github.com/scylladb/gocql/pull/278/commits/45d7cf0e84e250bb128c9e60a089a7b5b145cf73) is not the best name, `all code` should be replaced with something more specific

**Comment by illia-li** [comment]
> sorry for being so picky about commit messages but [move all code to internal/tests/serialization](https://github.com/scylladb/gocql/pull/278/commits/45d7cf0e84e250bb128c9e60a089a7b5b145cf73) is not the best name, `all code` should be replaced with something more specific

Is it better now?

File: internal/tests/serialization/set.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+						s.runUnmarshalTest("unmarshal**zero", t, unmarshal, val, unmarshalInZero)
+					}
+				}
+			})
+		}
+	})
+}
+
+// RunCorruptTest runs tests for cases when the function should an error.
+func (s Set) RunCorruptTest(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
```
Comment (dkropachev) [inline internal/tests/serialization/set.go:58]: In order to avoid confusiton let's have another structure for these tests or, even split it into two (for negative serialization and deserialization tests), since you already logically split them:
`Set` -> `PositiveSerializationSet`
`NegativeMarshalSet`
`NetativeUnmarhsalSet`

Feel free to suggest better names for them, or provide contr argument why we hsould not do it.

Comment (illia-li) [reply internal/tests/serialization/set.go:58]: :) It's more difficult than just writing code.

Comment (illia-li) [reply internal/tests/serialization/set.go]: Done

**Review by dkropachev — APPROVED** [review]
Let's wait for @sylwiaszunejko approve

**Review by sylwiaszunejko — APPROVED** [review]
